src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java

```diff
@@ -110,7 +110,8 @@ public void init() {
     return targetPullRequests;
   }
 
-  public String postBuildStartCommentTo(StashPullRequestResponseValue pullRequest) {
+  public String postBuildStartCommentTo(StashPullRequestResponseValue pullRequest)
+      throws StashApiException {
```

Comment:
do we need to add this to the signature?

Reply:
Yes, it's needed, and it's a good thing. The compiler is checking for me that the exception is either handled or passed to the caller.

This is not the right place to handle the exception. The caller knows more about the context and can write to the build log.

This indirection could be removed eventually. The `StashApiClient` object should not be owned by `StashRepository`. It should be available through `StashBuildTrigger` or through `StashCause` to all objects that deal with Bitbucket Server.

Reply:
OK, what got me confused is that we seem to be catching and handling some StashApiExceptions inside this code already?

Reply:
We handle `StashApiException` in that file, but not in that method.
`postBuildStartCommentTo()` should really be `private`, it is only called by `addFutureBuildTasks()` in the same class. `addFutureBuildTasks()` does the handling because it can do something meaningful. It skips building the PR, which is the right thing to do. If the plugin cannot post a comment indicating that the build starts, chances are it won't be able to post a comment about the build result, and then the PR would be considered eligible for building again, wasting Jenkins server resources.
As it stands now (i.e. without this PR), `RuntimeException` thrown by `StashApiClient` propagates all the way to the Jenkins core. The iteration over open pull requests stops when the exception is thrown.
If we handled exceptions in `postBuildStartCommentTo()`, we would need a way to indicate failure by other means, perhaps by returning `null` as the comment ID. We went through that in #93, it was your idea to use an exception instead of returning `null`.

Reply:
OK, thanks for the clarification.
Maybe it would be useful to add this explanation (or parts of) as comment/javadoc

Reply:
I'm trying to code self-explanatory. The compiler doesn't leave much freedom which checked exceptions I list on the interface, especially when combined with FindBugs/SpotBugs. The problem with long comments is that they become liabilities. They need to be edited as the code evolves, or they get out of date.

Reply:
Agree in general, though I think if you added the first paragraph next to the place where we catch the exception it could be easily maintained. The only change I expect is if we stop handling it there, but it's easy to remove the comment then.